### PR TITLE
fix: validate list `suffix` in n.add and n.remove and fix Series/DataFrame kwarg crash

### LINKED ISSUE #1650: Allow passing list in `n.add`/ `n.remove` to suffix
## Describe the feature you'd like to see
See #1629

### PR DESCRIPTION
**Pull request by darsh-mishr**
Closes #1650.

Two related issues with the `suffix` argument of `n.add` / `n.remove`. First, when both `name` and `suffix` were list-like of equal length, pandas broadcasting silently paired them element-by-element instead of erroring, so six buses and six carriers removed 6 components instead of the expected 36, with no warning. Second, combining a list `suffix` with a `pd.Series` / `pd.DataFrame` kwarg crashed with `TypeError: endswith first arg must be str or a tuple of str, not list`.

Per the design agreed on the issue thread (option 3), this allows `name` or `suffix` to be a list, but not both; list+list now raises a clear `ValueError`. The `pd.Series` / `pd.DataFrame` path is fixed by gating the suffix auto-rename to scalar suffix only. Both methods get the widened type signature and updated docstrings.

<details>
<summary>Reproducers and fix detail</summary>

```python
# Bug 1: silent zip in n.remove
n.remove("Generator", [f"DE0 {i}" for i in range(6)],
         suffix=[" onwind", " offwind-ac", " offwind-dc",
                 " solar-pv-utility", " solar-pv-rooftop", " ror"])
# before: 30 of 36 components silently remain
# after:  ValueError pointing at the two supported shapes

# Bug 2: TypeError on pd.Series kwarg with list suffix
n.add("Generator", "g", suffix=[" wind", " solar"],
      bus="bus", p_nom=pd.Series([100, 200], index=["g wind", "g solar"]))
# before: TypeError: endswith first arg must be str or a tuple of str, not list
# after:  created [g wind, g solar] with p_nom=[100, 200]
```

The changes in `pypsa/network/transform.py`:

- Widen `suffix` from `str` to `str | Sequence[str]` on both methods.
- Add a list+list guard before `names.astype(str) + suffix` that raises `ValueError`.
- Update `single_component` after the guard so a scalar `name` + list `suffix` is treated as multi-component for the kwarg loop.
- Gate the Series/DataFrame suffix auto-rename to scalar suffix; for list suffix the caller passes full post-suffix labels and the existing `v.index.equals(names)` check enforces alignment.

The design was confirmed by `@lkstrp` on the issue, referencing PR #1629's discussion with `@FabianHofmann` and `@euronion`.
</details>

## Checklist

- [x] Code changes are sufficiently documented.
- [x] Unit tests for the bug fix were added.
- [x] A note for the release notes `docs/release-notes.md` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

File: pypsa/network/transform.py [hunk 1]
```diff
@@ -204,10 +207,28 @@ def add(
...
+                f"Cannot pass list to both `name` and `suffix` for n.add of {class_name}; "
+                "the previous behaviour silently paired them one-by-one. Pass a list to "
+                "`name` with a scalar `suffix`, or a scalar `name` with a list `suffix`."
+            )
+            raise ValueError(msg)
+
         names = pd.Index([name]) if single_component else pd.Index(name)
+        # For scalar `name` + list `suffix`, broadcast the name to the suffix
+        # length so the addition is shape-matched under pandas 3 (which no
+        # longer auto-broadcasts length-1 to length-N). See #1650.
```
Comment (FabianHofmann) [inline pypsa/network/transform.py]: make this a one-liner comment or even better remove

File: pypsa/network/transform.py [hunk 2]
```diff
@@ -204,10 +207,28 @@ def add(
...
+        if single_component and suffix_is_list:
+            names = names.repeat(len(suffix))
         names = names.astype(str) + suffix
 
         names_str = "name" if single_component else "names"
+        # When `name` is scalar but `suffix` is a list, the result has multiple
+        # components; treat downstream kwargs as static (indexed by names) rather
+        # than time-varying (indexed by snapshots). See #1650.
```
Comment (FabianHofmann) [inline pypsa/network/transform.py]: same here

File: pypsa/network/transform.py [hunk 3]
```diff
@@ -255,21 +276,25 @@ def add(
                 if not v.index.equals(self.snapshots):
                     raise ValueError(msg.format(f"Series {k}", "network snapshots"))
             elif isinstance(v, pd.Series):
-                # Cast names index to string + suffix
-                v = v.rename(
-                    index=lambda s: (
-                        str(s) if str(s).endswith(suffix) else str(s) + suffix
+                # Cast names index to string + suffix (only when suffix is scalar;
+                # for list suffix the user is expected to pass the full post-suffix
+                # names directly, see #1650).
```
Comment (FabianHofmann) [inline pypsa/network/transform.py]: same here

File: pypsa/network/transform.py [hunk 4]
```diff
@@ -255,21 +276,25 @@ def add(
...
                     )
-                )
                 if not v.index.equals(names):
                     raise ValueError(msg.format(f"Series {k}", names_str))
             if isinstance(v, pd.DataFrame):
-                # Cast names columns to string + suffix
-                v = v.rename(
-                    columns=lambda s: (
-                        str(s) if str(s).endswith(suffix) else str(s) + suffix
+                # Cast names columns to string + suffix (scalar suffix only, see #1650).
```
Comment (FabianHofmann) [inline pypsa/network/transform.py]: same here

File: test/test_add_remove_suffix.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
+# SPDX-FileCopyrightText: PyPSA Contributors
```
Comment (FabianHofmann) [inline test/test_add_remove_suffix.py]: I think this is a bit overkill - a single test file for component removal with suffix. if these tests do not fit anywhere, rename this to a more general name to make it extendable later ie. `test_component_removal`

**Review by FabianHofmann — CHANGES_REQUESTED** [review]

File: pypsa/network/transform.py [hunk 6]
```diff
@@ -84,6 +84,41 @@ def is_typo(custom_attr: str, default_attr: str) -> bool:
...
+    single_component: bool,
+    op: str,
+) -> tuple[pd.Index, str | Sequence[str]]:
+    """Build the post-suffix component index for `n.add`/`n.remove`.
+
+    Validates list/list and empty-list suffixes, normalises a single-element
+    list suffix to a scalar so it behaves identically to the scalar form, and
+    broadcasts a scalar `name` across a list `suffix`.
+    """
+    suffix_is_list = not isinstance(suffix, str) and is_1d_list_like(suffix)
```
Comment (darsh-mishr) [inline pypsa/network/transform.py]: Minor optional thought: these three checks all gate on `suffix_is_list`, so nesting them under a single `if suffix_is_list:` reads a touch cleaner and evaluates the flag once. It also makes the message ordering explicit (a list `name` + empty-list `suffix` hits the `not single_component` branch first).

```python
suffix_is_list = not isinstance(suffix, str) and is_1d_list_like(suffix)
if suffix_is_list:
    if not single_component:
        raise ValueError(...)          # list + list
    if len(suffix) == 0:
        raise ValueError(...)          # empty list
    if len(suffix) == 1:               # normalise single-element to scalar
        suffix = next(iter(suffix))
        suffix_is_list = False
```

Purely stylistic, feel free to ignore.

Comment (FabianHofmann) [reply pypsa/network/transform.py]: yep! take another look

Comment (darsh-mishr) [reply pypsa/network/transform.py]: Looks good, and the `_append_suffix` extraction is a nice touch on top of the grouping. Reads cleanly now.

**Review by FabianHofmann — APPROVED** [review]
@darmis007 as you see I appened some change, this is good to go

**Comment by darsh-mishr** [comment]
Thanks a lot for taking this the rest of the way, Fabian. The helper extraction, the wider typing across the wrappers, and the extra edge-case tests made it much cleaner than my first pass, and I picked up a few things from your changes. Really appreciate the time you put into polishing it. Good to merge from my side.

**Review by lkstrp — APPROVED** [review]
